// Tests for the `joe` package's global functions

//-----------------------------------------------------------------------------
// catch()

function testCatch_ok() {
    var cr = catch(\-> "abc");
    assertT(cr.isOK());
    assertF(cr.isError());
    assertEQ(cr.result, "abc");
    assertEQ(cr.error, null);
}

function testCatch_error() {
    function bad() { throw "Simulated error!"; }
    var cr = catch(bad);
    assertF(cr.isOK());
    assertT(cr.isError());
    assertEQ(cr.result, null);
    assertEQ(cr.error.message(), "Simulated error!");
}

function testCatch_error_withTrace() {
    function bad() { throw "Simulated error!"; }
    var cr = catch(bad);
    assertF(cr.isOK());
    assertT(cr.isError());
    assertEQ(cr.result, null);
    assertEQ(cr.error.message(), "Simulated error!");

    assertEQ(cr.error.traces(), [
        "In function bad()",
        "In java call(<function bad()>)",
        "Called from catch()",
    ]);
}

